### PR TITLE
chore: improve typescript config and types

### PR DESCRIPTION
**Pull request by JustinBeckwith**
FYI @googleapis/yoshi-nodejs this is currently busted on master, and this *should* fix things.

### CODE REVIEW

File: test/codec.ts [hunk 1]
```diff
@@ -479,14 +477,12 @@ describe('codec', () => {
     it('should encode structs', () => {
       const value = codec.Struct.fromJSON({a: 'b', c: 'd'});
       const encoded = codec.encode(value);
-      assert.deepStrictEqual(encoded, ['b', 'd']);
+      assert.deepStrictEqual([].concat(encoded.slice()), ['b', 'd']);
```
Comment (JustinBeckwith) [inline test/codec.ts]: @ofrobots I am not wild about this test change, but I can't see how my change above would require it.  Can you spot something with my es6-ification of the struct?

Comment (ofrobots) [reply test/codec.ts]: Which change above are you referring to? How was the compiler complaining with the original form? Sorry, I am not familiar with the code here.

File: src/codec.ts [hunk 2]
```diff
@@ -84,73 +84,67 @@ const TYPE = Symbol();
...
-  return struct;
+  /**
+   * Checks to see if the provided object is a Struct.
+   *
+   * @private
+   *
+   * @param {*} thing The object to check.
+   * @returns {boolean}
+   */
+  static isStruct(thing: {}) {
```
Comment (ofrobots) [inline src/codec.ts]: Good opportunity for:
```suggestion
  static isStruct(thing: {}) : thing is Struct {
```

File: src/database.ts [hunk 3]
```diff
@@ -156,7 +167,7 @@ class Database extends ServiceObject {
       createMethod: (_, options, callback) => {
         return instance.createDatabase(formattedName_, options, callback);
       },
-    });
+    } as {} as ServiceObjectConfig);
```
Comment (kinwa91) [inline src/database.ts]: Curious why this is needed

Comment (kinwa91) [reply src/database.ts]: Ah, we don't have `requestModule` here. OK for now - we'll have to add types to `common-grpc`

**Review by jkwlui — APPROVED** [review]

**Pull request merged**